scripts/syntactics/parser/biaffine_parser.py

```diff
@@ -0,0 +1,370 @@
+
+            Parameters
+            ----------
+            seq_len : int
+                length of sequence
+            batch_size : int
+                batch size
+
+            Returns
+            -------
+            np.ndarray
+                dropout mask for word and tag
+            """
+            wm, tm = nd.zeros((seq_len, batch_size, 1)), nd.zeros((seq_len, batch_size, 1))
+            for i in range(seq_len):
+                word_mask = np.random.binomial(1, 1. - dropout_dim, batch_size).astype(np.float32)
+                tag_mask = np.random.binomial(1, 1. - dropout_dim, batch_size).astype(np.float32)
+                scale = 3. / (2. * word_mask + tag_mask + 1e-12)
+                word_mask *= scale
+                tag_mask *= scale
+                word_mask = nd.array(word_mask)
+                tag_mask = nd.array(tag_mask)
+                wm[i, :, 0] = word_mask
+                tm[i, :, 0] = tag_mask
+            return wm, tm
+
+        self.generate_emb_mask = _emb_mask_generator
+
+    def parameter_from_numpy(self, name, array):
+        """ Create parameter with its value initialized according to a numpy tensor
+
+        Parameters
+        ----------
+        name : str
+            parameter name
+        array : np.ndarray
+            initiation value
+
+        Returns
+        -------
+        mxnet.gluon.parameter
+            a parameter object
+        """
+        p = self.params.get(name, shape=array.shape, init=mx.init.Constant(array))
+        return p
+
+    def parameter_init(self, name, shape, init):
+        """Create parameter given name, shape and initiator
+
+        Parameters
+        ----------
+        name : str
+            parameter name
+        shape : tuple
+            parameter shape
+        init : mxnet.initializer
+            an initializer
+
+        Returns
+        -------
+        mxnet.gluon.parameter
+            a parameter object
+        """
+        p = self.params.get(name, shape=shape, init=init)
+        return p
+
+    def forward(self, word_inputs, tag_inputs, arc_targets=None, rel_targets=None):
+        """Run decoding
+
+        Parameters
+        ----------
+        word_inputs : mxnet.ndarray.NDArray
+            word indices of seq_len x batch_size
+        tag_inputs : mxnet.ndarray.NDArray
+            tag indices of seq_len x batch_size
+        arc_targets : mxnet.ndarray.NDArray
+            gold arc indices of seq_len x batch_size
+        rel_targets : mxnet.ndarray.NDArray
+            gold rel indices of seq_len x batch_size
+        Returns
+        -------
+        tuple
+            (arc_accuracy, rel_accuracy, overall_accuracy, loss) when training, else if given gold target
+        then return arc_accuracy, rel_accuracy, overall_accuracy, outputs, otherwise return outputs, where outputs is a
+        list of (arcs, rels).
+        """
+        is_train = autograd.is_training()
```

Comment:
I suggest moving this logic to the main training/inference loops.

Reply:
Yes, done.